### ISSUE TITLE
Problem with using emulator

### ISSUE THREAD

**Issue by bimusiek** [thyrlian/AndroidSDK#1]
Hey, I am trying all day different configurations and I cannot really make it happen. 

When I try to install any .apk I get an error:
```
root@a717690b4c77:/workspace# adb install /opt/android-sdk/system-images/android-26/google_apis/x86/data/app/SmokeTestApp/SmokeTestApp.apk 
Failed to install /opt/android-sdk/system-images/android-26/google_apis/x86/data/app/SmokeTestApp/SmokeTestApp.apk: Can't find service: package
```

Listing services of emulator indeed does not return package service and I don't know why.
```
adb shell service list
Found 6 services:
0	gpu: [android.ui.IGpuService]
1	SurfaceFlinger: [android.ui.ISurfaceComposer]
2	android.service.gatekeeper.IGateKeeperService: []
3	android.security.keystore: [android.security.IKeystoreService]
4	android.hardware.fingerprint.IFingerprintDaemon: []
5	batteryproperties: [android.os.IBatteryPropertiesRegistrar]
```

Any ideas?

My `Dockerfile`:
```
FROM thyrlian/android-sdk
RUN sdkmanager --update
RUN sdkmanager "system-images;android-24;google_apis;armeabi-v7a" --verbose
RUN echo "no" | avdmanager create avd -n test -k "system-images;android-24;google_apis;armeabi-v7a" -b default/armeabi-v7a -f
```

**Comment by thyrlian** [comment]
Hi, I just tried with your steps, but can't reproduce it.

According to your `Dockerfile`, you have:

```
sdkmanager "system-images;android-24;google_apis;armeabi-v7a" --verbose
```

and

```
echo "no" | avdmanager create avd -n test -k "system-images;android-24;google_apis;armeabi-v7a" -b default/armeabi-v7a -f
```

Actually there is a problem:

The image you downloaded is `system-images;android-24;google_apis;armeabi-v7a`, while you specified `-b default/armeabi-v7a` during avd creation.  That's inconsistent, which would encounter such an error:

```
Error: Invalid --tag default for the selected package. Valid tags are:
google_apis
```

What exactly is `-b` parameter for `avdmanager create`?

```
-b --abi     : The ABI to use for the AVD. The default is to auto-select the ABI if the platform has only one ABI for its system images.
```

Available emulator system images to download:

```
system-images;android-24;default;armeabi-v7a
system-images;android-24;google_apis;armeabi-v7a
```

You should try:

```
echo "no" | avdmanager create avd -n test -k "system-images;android-24;google_apis;armeabi-v7a" -b google_apis/armeabi-v7a -f
```

or download `system-images;android-24;default;armeabi-v7a` image and try:

```
echo "no" | avdmanager create avd -n test -k "system-images;android-24;default;armeabi-v7a" -b default/armeabi-v7a -f
```

Once the avd is created successfully, you should be able to see it by:

```
avdmanager list avd
```

And then just run the emulator like this:

```
emulator64-arm -avd test -noaudio -no-boot-anim -no-window -accel on &
```

```
adb devices
List of devices attached
emulator-5554	device

adb shell service list
Found 21 services:
0	display: [android.hardware.display.IDisplayManager]
1	power: [android.os.IPowerManager]
2	appops: [com.android.internal.app.IAppOpsService]
3	batterystats: [com.android.internal.app.IBatteryStats]
4	media.camera: [android.hardware.ICameraService]
5	media.sound_trigger_hw: [android.hardware.ISoundTriggerHwService]
6	media.radio: [android.hardware.IRadioService]
7	media.audio_policy: [android.media.IAudioPolicyService]
8	media.resource_manager: [android.media.IResourceManagerService]
9	media.player: [android.media.IMediaPlayerService]
10	media.drm: [android.media.IMediaDrmService]
11	media.extractor: [android.media.IMediaExtractorService]
12	media.audio_flinger: [android.media.IAudioFlinger]
13	drm.drmManager: [drm.IDrmManagerService]
14	media.codec: [android.media.IMediaCodecService]
15	gpu: [android.ui.IGpuService]
16	SurfaceFlinger: [android.ui.ISurfaceComposer]
17	android.service.gatekeeper.IGateKeeperService: []
18	android.security.keystore: [android.security.IKeystoreService]
19	android.hardware.fingerprint.IFingerprintDaemon: []
20	batteryproperties: [android.os.IBatteryPropertiesRegistrar]
```

Finally, I tried to download an APK here:

```
wget https://github.com/emmaguy/clean-status-bar/releases/download/1.1.4/app-release.apk
```

Then install it:

```
adb install app-release.apk
Success
```

Check the installed packages:

```
adb shell pm list packages | grep cleanstatusbar
package:com.emmaguy.cleanstatusbar
```

**Comment by bimusiek** [comment]
I tried many different variations and still could not install APK :( 

I work on OSX, here are my steps:
```
docker run -it thyrlian/android-sdk:latest bash
#then inside container
sdkmanager --update
sdkmanager "system-images;android-24;default;armeabi-v7a" "platforms;android-24" --verbose
echo "no" | avdmanager create avd -n test -k "system-images;android-24;default;armeabi-v7a" -b default/armeabi-v7a -f
emulator64-arm -avd test -noaudio -no-boot-anim -no-window -accel on &
```

List of services
```
root@5dff1d190087:/opt/android-sdk/tools# adb shell service list
Found 17 services:
0	media.camera: [android.hardware.ICameraService]
1	media.sound_trigger_hw: [android.hardware.ISoundTriggerHwService]
2	media.radio: [android.hardware.IRadioService]
3	media.audio_policy: [android.media.IAudioPolicyService]
4	media.resource_manager: [android.media.IResourceManagerService]
5	media.player: [android.media.IMediaPlayerService]
6	media.drm: [android.media.IMediaDrmService]
7	media.audio_flinger: [android.media.IAudioFlinger]
8	media.extractor: [android.media.IMediaExtractorService]
9	drm.drmManager: [drm.IDrmManagerService]
10	media.codec: [android.media.IMediaCodecService]
11	gpu: [android.ui.IGpuService]
12	SurfaceFlinger: [android.ui.ISurfaceComposer]
13	android.hardware.fingerprint.IFingerprintDaemon: []
14	android.security.keystore: [android.security.IKeystoreService]
15	android.service.gatekeeper.IGateKeeperService: []
16	batteryproperties: [android.os.IBatteryPropertiesRegistrar]
```

And trying to install .apk
```
root@5dff1d190087:/tmp# adb install app-release.apk
Failed to install app-release.apk: Can't find service: package
```

I have tried on Ubuntu, the same issue.

Can it be that I do `sdkmanager --update` directly in Docker container?

**Comment by thyrlian** [comment]
Yep, that could be the issue.  As I mentioned in [README](https://github.com/thyrlian/AndroidSDK#caveat), the default FS used by Docker is AUFS, which causes failure during upgrading Android SDK.

But you can still try to update Android SDK inside a Docker container.  The trick is when you spin up a container, you have to mount an external SDK volume from host to the container, just like below:

```console
# Please replace the host machine volume directory with your preference
# There are two placeholders below: $(pwd)/sdk

# First copy the default Android SDK to the host machine
docker stop $(docker ps -aqf "ancestor=thyrlian/android-sdk") &> /dev/null && docker rm $(docker ps -aqf "ancestor=thyrlian/android-sdk") &> /dev/null; docker run -it --rm -v $(pwd)/sdk:/sdk thyrlian/android-sdk:sdk:latest bash -c 'cp -a $ANDROID_HOME/. /sdk'

# Run a container with the Android SDK volume mounted
docker run -it -v $(pwd)/sdk:/opt/android-sdk thyrlian/android-sdk:latest bash
```

That should be it.  Please try it out again and share your results.  Thanks and good luck.

**Comment by bimusiek** [comment]
I will try. Thanks a lot @thyrlian ! I owe you a beer 👍

**Comment by thyrlian** [comment]
No problem @bimusiek , glad to help.  Please post your results here after trying, I'm curious to know it.  BTW, maybe we could also try to meet in person in Berlin, I'm interested in your twin entrepreneur :)

**Comment by bimusiek** [comment]
I just tried to run it as you suggested:
1. Copied sdk from your docker image to local folder
2. Started container with volume `./web/sdk/:/opt/android-sdk`
3. In container I have run
```

sdkmanager --update --verbose
sdkmanager "system-images;android-24;google_apis;armeabi-v7a" "platforms;android-24" --verbose
echo "no" | /opt/android-sdk/tools/bin/avdmanager create avd -n test -k "system-images;android-24;google_apis;armeabi-v7a" -b google_apis/armeabi-v7a -f

adb start-server
emulator64-arm -avd test -noaudio -no-boot-anim -no-window -accel on -gpu off &

wget https://github.com/emmaguy/clean-status-bar/releases/download/1.1.4/app-release.apk
adb install app-release.apk
```

And at the end the same issue:
```
root@561e2fba3bde:/workspace# adb install app-release.apk
Failed to install app-release.apk: Can't find service: package
```

I have even installed Ubuntu on Parallel and directly in the console I followed all steps from your `Dockerfile` and the same results 😢  

APK is not installable. I am slowly loosing hope.

Basically, what we are trying to do is start android emulator on Google Cloud in Kubernetes with APK that has server inside to be able to use Google Mobile Vision API. APK is ready with endpoint, I just cannot figure out how to start it... 

Do you think you could create docker image with emulator inside and I would try to pull it? 
I will try to run everything now on Digitalocean instance to see if it helps, but yesterday I had the same problem. 

Or maybe you could send me zipped android-sdk you have to mike@ahoy.io and I can see if this works? 

About Berlin, sure, let's do it! Just drop me an email when and where, I am quite flexible 👍

**Comment by thyrlian** [comment]
Your approach looks all good, I just don't get it why it fails.

The zipped Android SDK on my machine is the same one downloaded from [here](https://dl.google.com/android/repository/sdk-tools-linux-3859397.zip).

I actually tried on my macOS with exact the same steps and succeeded.  Would you mind sending me any temporary remote ssh username / password so that we can see together if it works on your environment?  My email address is: thyrlian@gmail.com

**Comment by bimusiek** [comment]
I tried now on 3 different servers 🙈 Same issue.

But... I think we found the problem... My friend was testing on his OSX, then he got the same result. 
He was curious so he run shell services again and it showed more than 100 and he was able to install apk.
So it looks like emulator needs some time to start everything! 
Trying to configure it with our APK now to see if it works.

**Comment by bimusiek** [comment]
Yes, that was it. I have to wait 2-3mins until emulator is up and running. 
Omg, I just wasted 1,5 day tracking the issue.

**Comment by thyrlian** [comment]
You're right.  That makes sense - I just tested, trying `adb install` when device is still booting (launching necessary services), the same error occurs `Can't find service: package`.

Sorry that I thought you had waited for the emulator is ready.

* This docker image is mainly for Android CI / compiling.  You can use emulator inside, but it's not recommended.
* ARM emulator is slow, X86 is preferred, but that needs a physical Linux machine with KVM enabled.
* If you want to have some pipeline job which consume the emulator, you'd better write a helper script to detect the readiness / health status of the emulator (e.g. the launch waiting time).
* I would recommend 3rd party service like Genymotion cloud.

**Comment by thyrlian** [comment]
Cool, congratulations for solving the problem :)

**Comment by bimusiek** [comment]
Thanks for tips, it is experimental for now and we can easily scale Google Cloud to just spawn more containers. I will be testing how slow is the endpoint when it is deployed 👍

**Comment by thyrlian** [comment]
Have fun 👍   So I'll email you in the future to meet you in Berlin!
